Conflict-resolution-perf-improvements.md

```diff
@@ -0,0 +1,112 @@
+  - Where `X` could be configurable, but should probably just be an internal constant
+
+* Each node contains:
+    - A Map to store leaf nodes, which enables amortised constant lookup and to accomodate `.hashcode()` conflicts on keys
+    - A node hash field containing a hash of all the leaf node's hashes combined
+
+* The tree, excluding leafs, can be represented as a int and int[X] (Root+X hashes) in a RPC
+
+* During CR if two tree's root values don't match, it's then possible to compare all `X` hashes and only request the indexes
+of the array which have conflicting hashes. At which point the participating nodes only return the InternalCacheEntries
+associated with those nodes.
+
+* Once the hash tree has been created for CR, it's possible to cache the hashes until a subsequent write operation. Upon
+a write operation it's necessary to invalidate the root hash and the hash of the child node containing the entry. During
+the next round of CR it's then only necessary to recalculate the invalidated nodes and the root.
+
+* Logic for processing write operations can be implemented as an interceptor which interacts with the conflict manager.
+
+### Offline Backup Consistency Check
+The Merkle tree hashes can also be utilised to ensure the consistency of data being dumped to an offline backup. When a
+user initiates a backup:
+
+1. The cluster initiates CR to ensure that no inconsistencies exist before the backup
+2. A new Merkle tree is created, with the root node being the hash of all primary replica segment hashes
+3. The root hash is stored as part of the backup metadata
+4. When a cluster is restored from a backup, the clusterwide Merkle tree can be recreated and the new root hash can be
+compared to the value stored in the metadata.
+
+NOTE: Utilising Merkle trees in this manner would mean that it's not possible to make changes to how entries are hashed
+and how the tree is created without an appropriate migration strategy between Infinispan versions.
+
+## Distributed CR processing
+Instead of CR being coordinated by a single node, the coordinator should instruct the primary owner of each segment to
+initiate CR. As it's likely that a single node will be the primary for multiple segments in a small cluster, we should
+still limit each node to executing CR for a single segment at a time.
+
+Distributing CR also benefits from the use of Merkle trees, as it means that the primary owner in the preferred partition
+who is coordinating the CR would not have to send their tree over the wire.
+
+TRACKED BY: [ISPN-9084] https://issues.redhat.com/browse/ISPN-9084
+
+## Prioritise segments based upon requests during merge
+During a partition merge, if CR is in progress and a request is made on a specific key before it's segment has been processed,
+then we should perform CR on the key in place.
+
+The advantage is that conflicts on an active entitiy are resolved sooner, furthemore, if (## Maintain a segment hash) is
```

Comment:
I think it's `<<Maintain a segment hash>>`
https://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#links

Reply:
Sorry, I thought this was asciidoc, but I see now it's markdown :)
https://stackoverflow.com/questions/2822089/how-to-link-to-part-of-the-same-document-in-markdown/16426829#16426829

Reply:
I just blindly changed haha